odl/trafos/wavelet.py

```diff
@@ -834,14 +957,511 @@ def adjoint(self):
         if self.is_orthogonal:
             return self.inverse
         else:
-            # TODO: put adjoint here
-            return super().adjoint
+            if not isinstance(self.wbasis, pywt.Wavelet):
+                return InverseAdjBiorthWaveletTransform(dom=self.range,
+                                                        nscales=self.nscales,
+                                                        wbasis=self.wbasis)
+            else:
+                return super().adjoint
 
     @property
     def inverse(self):
         """The inverse wavelet transform."""
-        return WaveletTransform(dom=self.range, nscales=self.nscales,
-                                wbasis=self.wbasis, mode=self.mode)
+        if not isinstance(self.wbasis, pywt.Wavelet):
+            return BiorthWaveletTransform(dom=self.range, ran=self.domain,
+                                          nscales=self.nscales,
+                                          wbasis=self.wbasis)
+        elif isinstance(self.wbasis, pywt.Wavelet):
+            return WaveletTransform(dom=self.range, nscales=self.nscales,
+                                    wbasis=self.wbasis, mode=self.mode)
+
+
+class BiorthWaveletTransform(Operator):
```

Comment:
The separate `BiorthWaveletTransform` is due to the Jan-Olov's wavelet library.
The biorthogonal wavelets that are in pyWavelets have even filter lengths, Jan-Olov's wavelets have odd filter length. Also for Jan-Olov's wavelets there is own classes for the adjoint transform and the inverse adjoint transform, these we don't have for biorthogonal pyWavelets (with orthogonal wavelets they are not necessary). _Maybe_ Jan-Olov will write adjoint and inverse adjoint codes for even length filters at some point, but that remains to be seen...

I'll add some comment to the code related to this